### ISSUE TITLE
Byte based scaling stucks with Reader already checkpointed exception

### ISSUE THREAD

**Issue by Raghuveera-Rao** [pravega/pravega#3837]
While trying Byte based with samples, stalls with **Exception in thread "Thread-20" java.lang.IllegalStateException: Reader already checkpointed**  after first 60 mins of segment scale. Scale configurations are tried with scaling.policy=bytesByRate ,10 writer, 10 reader, min segment 1 and scale factor 2.

Pravega details:
```
Pravega version : 0.5.0-2252.c180711
Zookeeper Operator : pravega/zookeeper-operator:0.2.1
Pravega Operator: pravega/pravega-operator:0.3.
```

```
root@pravega-benchmark:/pravega-puncher/scripts# ./runFatByteScaler | grep ===
16:07:44.866 [pool-2-thread-1] INFO io.pravega.puncher.streams.scaling.ScalingCase - ============ segments: 1
16:19:44.847 [pool-2-thread-1] INFO io.pravega.puncher.streams.scaling.ScalingCase - ============ segments: 2
16:29:44.844 [pool-2-thread-1] INFO io.pravega.puncher.streams.scaling.ScalingCase - ============ segments: 4
16:39:44.844 [pool-2-thread-1] INFO io.pravega.puncher.streams.scaling.ScalingCase - ============ segments: 8
16:49:44.843 [pool-2-thread-1] INFO io.pravega.puncher.streams.scaling.ScalingCase - ============ segments: 10
Exception in thread "Thread-20" java.lang.IllegalStateException: Reader already checkpointed.
        at com.google.common.base.Preconditions.checkState(Preconditions.java:507)
        at io.pravega.client.stream.impl.CheckpointState.readerCheckpointed(CheckpointState.java:116)
        at io.pravega.client.stream.impl.ReaderGroupState$CheckpointReader.update(ReaderGroupState.java:843)
        at io.pravega.client.stream.impl.ReaderGroupState$ReaderGroupStateUpdate.applyTo(ReaderGroupState.java:452)
        at io.pravega.client.stream.impl.ReaderGroupState$ReaderGroupStateUpdate.applyTo(ReaderGroupState.java:446)
        at io.pravega.client.state.impl.StateSynchronizerImpl.applyUpdates(StateSynchronizerImpl.java:127)
        at io.pravega.client.state.impl.StateSynchronizerImpl.conditionallyWrite(StateSynchronizerImpl.java:231)
        at io.pravega.client.state.impl.StateSynchronizerImpl.updateState(StateSynchronizerImpl.java:135)
        at io.pravega.client.stream.impl.ReaderGroupStateManager.checkpoint(ReaderGroupStateManager.java:411)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.updateGroupStateIfNeeded(EventStreamReaderImpl.java:170)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEventInternal(EventStreamReaderImpl.java:105)
        at io.pravega.client.stream.impl.EventStreamReaderImpl.readNextEvent(EventStreamReaderImpl.java:90)
        at io.pravega.puncher.streams.scaling.ScaledEventReader.run(ScaledEventReader.java:41)
        at java.lang.Thread.run(Thread.java:748)
```

Segment stores are showing following exceptions on these streams repeatedly:
```
2019-05-24 16:49:44,997 22572978 [epollEventLoopGroup-11-6] INFO  i.p.s.s.h.handler.AppendProcessor - Setting up appends for writer: 3fef87aa-8abd-48ce-bcc9-2dbcf3619936 on segment: puncherScope1/puncherStream1/4.#epoch.2
2019-05-24 16:49:45,030 22573011 [epollEventLoopGroup-11-6] ERROR i.p.s.s.h.h.ServerConnectionInboundHandler - Caught exception on connection: 
io.netty.handler.codec.DecoderException: io.pravega.shared.protocol.netty.InvalidMessageException: AppendBlockEnd without AppendBlock.
    at io.netty.handler.codec.MessageToMessageDecoder.channelRead(MessageToMessageDecoder.java:98)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:362)
    at io.netty.channel.AbstractChannelHandlerContext.invokeChannelRead(AbstractChannelHandlerContext.java:348)
    at io.netty.channel.AbstractChannelHandlerContext.fireChannelRead(AbstractChannelHandlerContext.java:340)
    at io.netty.handler.codec.ByteToMessageDecoder.fireChannelRead(ByteToMessageDecoder.java:323)
    at io.netty.handler.codec.ByteToMessageDecoder.channelRead(ByteToMessageDecoder.java:297)
```

This is seen in controller logs:
```
2019-05-24 16:50:17,524 22666067 [controllerpool-4] ERROR i.p.c.e.i.ConcurrentEventProcessor - ConcurrentEventProcessor Processing failed, exiting {}
java.util.concurrent.CompletionException: io.pravega.controller.store.stream.EpochTransitionOperationExceptions$PreConditionFailureException
    at java.util.concurrent.CompletableFuture.encodeRelay(CompletableFuture.java:326)
    at java.util.concurrent.CompletableFuture.completeRelay(CompletableFuture.java:338)
    at java.util.concurrent.CompletableFuture.uniRelay(CompletableFuture.java:911)
    at java.util.concurrent.CompletableFuture$UniRelay.tryFire(CompletableFuture.java:899)
    at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:474)
    at java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:1977)
    at io.pravega.controller.server.eventProcessor.requesthandlers.ScaleOperationTask.lambda$execute$0(ScaleOperationTask.java:73)
    at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:760)
    at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:736)
    at java.util.concurrent.CompletableFuture$Completion.run(CompletableFuture.java:442)
    at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
    at java.util.concurrent.FutureTask.run(FutureTask.java:266)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:180)
    at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:293)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
    at java.lang.Thread.run(Thread.java:748)
```

**Comment by fpj** [comment]
@Raghuveera-Rao you are using a stale version of Pravega that includes a fix for the decoder issue.

**Comment by Raghuveera-Rao** [comment]
> @Raghuveera-Rao you are using a stale version of Pravega that does not includes a fix for the decoder issue.

Will try with 0.5.0-2260.b477eec and let you know.. Thanks..

**Comment by fpj** [comment]
Focusing on this exception:

```
Exception in thread "Thread-20" java.lang.IllegalStateException: Reader already checkpointed.
```

This exception is being raised as part of checking whether a particular reader has already been checkpoint or not. Specifically, whether the reader group state contains that information. It is a valid scenario that:

1- The reader tries to update the reader group state
2- The state is update but say the connection drops
3- The reader reconnects and tries again, leading to illegal state message

The fact that this is being reported as an error is only a matter of code and reporting style. This is not an issue.

**Comment by shrids** [comment]
Went through the logs, the exception 
```
java.lang.IllegalStateException: Reader already checkpointed.
```
is not an issue.
The DecoderException has already been resolved in the latest version of 0.5.0. 
Please re-open the issue if you observe it again.
